# Parsing
The **obejctive** of this lab is to to assign a syntactic structure to sentences by training and evaluating our: 
- constituent parsers (using PTB "Penn Tree Bank")
- dependency parsers (using UD "Universal Dependency")

The parse trees we are generating can be used in applications like grammar checkers.

## Constituent parsing (Berkeley parser)
### INSTALLATION
**Constituent parsing** is when you divide a sentence into constituents (smaller phrases that belong to a specific grammar category). 

Make sure you're in the right directory before downloading!

In [1]:
# current directory
!pwd

/Users/annabellepurnomo


In [2]:
# moving to the folder I want to download my files into
%cd /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/

/Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2


First we will download Berkeley Parser, a parsing algorithm written by Slav Petrov & Dan Klein (2007).

In [4]:
!git clone https://github.com/slavpetrov/berkeleyparser.git

Cloning into 'berkeleyparser'...
remote: Enumerating objects: 1399, done.
remote: Total 1399 (delta 0), reused 0 (delta 0), pack-reused 1399
Receiving objects: 100% (1399/1399), 90.05 MiB | 2.83 MiB/s, done.
Resolving deltas: 100% (756/756), done.
Updating files: 100% (331/331), done.


Open the Berkeley Parser tarball (bundle of files, non-compressed) to get its contents.

In [ ]:
!tar xvfz /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/UD_English-EWT.tar.gz

``cd`` (change directory/move) to our berkeleyparser folder 

In [2]:
%cd /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/berkeleyparser

/Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/berkeleyparser


### TRAINING
To generate our trees, we input an annotated parsed text from Penn Tree Bank called "02-21.10way_1Kclean" which is a small file that contains 1K lines.  

When training our language model with berkeley parser, it will do 6 iterations of  
- **split** randomly to get CFG rules  
- **merge** when the generated rules don't improve our parser 
- **smooth** to normalize our results 

There are intermediate files named "ptb.gr_*_merging.gr" generated after each iteration.

We output our final grammar as a binary file called "ptb.gr". 

First, let's learn grammar from trees contained in a SINGLEFILE.

In [74]:
!java -cp BerkeleyParser-1.7.jar edu.berkeley.nlp.PCFGLA.GrammarTrainer -path /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/02-21.10way_1K.clean -out ptb.gr -treebank SINGLEFILE

Calling with { -path => /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/02-21.10way_1K.clean -out => ptb.gr -treebank => SINGLEFILE }
Loading trees from /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/02-21.10way_1K.clean and using language SINGLEFILE
Will remove sentences with more than 10000 words.
Using horizontal=0 and vertical=1 markovization.
Using RIGHT binarization.
Using a randomness value of 1.0
Using grammar output file ptb.gr.
Random number generator seeded at 2.
I will do at least 50 iterations.
Using smoothing parameters 0.5 and 0.1
Loading data from single file!
Loading trees from single file...done
In training set we have # of words: 24358
reducing number of training trees from 1000 to 1000
Binarizing and annotating trees...done.
Binarizing and annotating trees...done.
There are 1000 tre

### PARSING 
We specify our newly generated grammar with "-gr ptb.gr" and parse the file "23.auto.clean" to output a parsed file called "23.auto.clean.parsed". 

In [3]:
!java -jar berkeleyParser-1.7.jar -gr ptb.gr < /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/23.auto.clean.txt > 23.auto.clean.parsed

This is what the first few lines of ``23.auto.clean.txt`` (original text file we parsed) file looks like. Note that it's already been tokenized (broken down) for us (we can tell because contracted words are split into two parts, "was" + "the clitic "n't", and punctuation is also separated by spaces)

In [4]:
!head /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/23.auto.clean.txt

No , it was n't Black Monday . 
But while the New York Stock Exchange did n't fall apart Friday as the Dow Jones Industrial Average plunged 190.58 points -- most of it in the final hour -- it barely managed to stay this side of chaos . 
Some `` circuit breakers '' installed after the October 1987 crash failed their first test , traders say , unable to cool the selling panic in both stocks and futures . 
The 49 stock specialist firms on the Big Board floor -- the buyers and sellers of last resort who were criticized after the 1987 crash -- once again could n't handle the selling pressure . 
Big investment banks refused to step up to the plate to support the beleaguered floor traders by buying big blocks of stock , traders say . 
Heavy selling of Standard & Poor 's 500-stock index futures in Chicago relentlessly beat stocks downward . 
Seven Big Board stocks -- UAL , AMR , BankAmerica , Walt Disney , Capital Cities\/ABC , Philip Morris and Pacific Telesis Group -- stopped trading and nev

Here is an output of our newly parsed file ``23.auto.clean.parsed``.

In [5]:
!head /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/berkeleyparser/23.auto.clean.parsed

( (S (ADVP (DT No)) (, ,) (NP (PRP it)) (VP (VBD was) (RB n't) (NP (NNP Black) (NNP Monday))) (. .)) )
( (S (CC But) (SBAR (IN while) (S (NP (DT the) (NNP New) (NNP York) (NNP Stock) (NNP Exchange)) (VP (VBD did) (RB n't) (NP (NP (NN fall)) (ADVP (RB apart)) (VP (VP (VBN Friday) (PP (IN as) (NP (DT the) (NNP Dow) (NNP Jones) (NNP Industrial) (JJ Average) (JJ plunged) (CD 190.58) (NNS points)))) (PRN (: --) (NP (NP (JJS most)) (PP (IN of) (NP (PRP it))) (PP (IN in) (NP (DT the) (JJ final) (NN hour)))) (: --))))))) (S (NP (PRP it)) (ADVP (RB barely)) (VP (VBZ managed) (S (VP (TO to) (VP (VB stay) (NP (NP (DT this) (NN side)) (PP (IN of) (NP (NNS chaos))))))))) (. .)) )
( (S (NP (NP (DT Some)) (SBAR (S (`` ``) (NP (NN circuit) (NNS breakers)) ('' '') (VP (VBD installed) (PP (IN after) (NP (DT the) (NNP October) (CD 1987) (NN crash))))))) (VP (VBD failed) (NP (PRP$ their) (JJ first) (NN test) (PRN (, ,) (S (NP (NNS traders)) (VP (VBP say))) (, ,)) (NN unable) (S (VP (TO to) (VP (VB cool) (

### EVALUATION
We use EVALB to evaluate see how accurate the bracketing in our test file (what we just parsed) to a gold file (a file that has all the correct answers). 

EVALB returns recall, precision, and tagging accuracy.

**Recall**: What proportion of actual positives was identified correctly?

**Precision**: What proportion of positive identifications was actually correct?

**Fmeasure**: Calculated using the results of precison and recall. 

Download EVALB.

In [8]:
!wget https://nlp.cs.nyu.edu/evalb/EVALB.tgz

--2021-10-14 16:42:28--  https://nlp.cs.nyu.edu/evalb/EVALB.tgz
Resolving nlp.cs.nyu.edu (nlp.cs.nyu.edu)... 216.165.22.203
Connecting to nlp.cs.nyu.edu (nlp.cs.nyu.edu)|216.165.22.203|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 38394 (37K) [application/x-gzip]
Saving to: ‘EVALB.tgz’

EVALB.tgz           100%[===================>]  37.49K  --.-KB/s    in 0.1s    

2021-10-14 16:42:29 (304 KB/s) - ‘EVALB.tgz’ saved [38394/38394]



Like when downloading Berkeley Parser, open the tarball to get the files from what we just downloaded.

In [11]:
!tar xvfz EVALB.tgz

x EVALB/
x EVALB/sample/
x EVALB/sample/sample.prm
x EVALB/sample/sample.gld
x EVALB/sample/sample.tst
x EVALB/sample/sample.rsl
x EVALB/Makefile
x EVALB/README
x EVALB/evalb
x EVALB/tgrep_proc.prl
x EVALB/LICENSE
x EVALB/new.prm
x EVALB/bug/
x EVALB/bug/bug.tst
x EVALB/bug/bug.rsl-new
x EVALB/bug/bug.rsl-old
x EVALB/bug/bug.gld
x EVALB/evalb.c
x EVALB/COLLINS.prm


``cd`` into EVALB.

In [6]:
%cd /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/berkeleyparser/EVALB 

/Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/berkeleyparser/EVALB


Download our gold file called ``23.auto.clean``.

In [81]:
!wget https://raw.githubusercontent.com/jhcross/span-parser/master/data/23.auto.clean

--2021-10-15 09:46:01--  https://raw.githubusercontent.com/jhcross/span-parser/master/data/23.auto.clean
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.110.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 845041 (825K) [text/plain]
Saving to: ‘23.auto.clean’

23.auto.clean       100%[===================>] 825.24K  --.-KB/s    in 0.09s   

2021-10-15 09:46:02 (9.39 MB/s) - ‘23.auto.clean’ saved [845041/845041]



### EVALB scorer format:

_evalb -p Parameter_file Gold_file Test_file_

**gold file:**
23.auto.clean 

**test file:** 
23.auto.clean.parsed

Using EVALB, and specifying the parameter (``-p``) as COLLINS.prm (the parameter sets options regarding the scoring method based on Collins (1997):
- make sure a constituent has the same span and label as the constituent in the gold file
- don't include the label on the top (the highest label)
- remove unneeded information logged/written into the trees called 'traces'
- remove certain punctuation
- don't include traces when calculating the length of a sentence)

Evaluating.

In [7]:
!./evalb -p COLLINS.prm 23.auto.clean ../23.auto.clean.parsed

  Sent.                        Matched  Bracket   Cross        Correct Tag
 ID  Len.  Stat. Recal  Prec.  Bracket gold test Bracket Words  Tags Accracy
   1    8    0   80.00  66.67     4      5    6      0      8     7    87.50
   2   40    0   74.19  74.19    23     31   31      3     40    34    85.00
   3   31    0   80.95  73.91    17     21   23      2     31    28    90.32
   4   35    0   66.67  51.85    14     21   27      5     35    30    85.71
   5   26    0   91.30  80.77    21     23   26      2     26    24    92.31
   6   17    0    8.33   8.33     1     12   12      6     17    12    70.59
   7   30    2    0.00   0.00     0      0    0      0      0     0     0.00
   8    6    0  100.00  83.33     5      5    6      0      6     5    83.33
   9    7    0   75.00  60.00     3      4    5      0      7     6    85.71
  10   36    0   95.65  84.62    22     23   26      0     36    35    97.22
  11   25    0   63.64  60.87    14     22   23      1     25    24    96.00
 

#### From these results, we see

for all sentences:

- Bracketing Recall         =  74.91
- Bracketing Precision      =  70.36
- Bracketing FMeasure       =  72.56

for sentences that have forty, or less than forty words:

- Bracketing Recall         =  74.91
- Bracketing Precision      =  70.36
- Bracketing FMeasure       =  72.56

We will just focus on sentences with less than forty words.

A perfect FMeasure (100) would indicate perfect precison and recall. With a score of 72.56, we understand that our parser is 72.56 in its recall and precison. Both our **recall** and **precison** are around 70-74, meaning our algorithm returns 70% relevant results than irrelevant ones (precison) and that our algorhitm returns 74% of the relevant results (recall). 

These are both precise and able to recall to a similar degree.

## Dependency parsing (UDPipe)
**Dependency Parsing:** links words together in a sentence to show relationships between words and their constitutes

We can use UDPipe, which is a Universal Dependency Pipeline toolkit that can help with 
- tokenization
- tagging
- lemmatization
- dependency parsing

We ``mkdir`` (make a directory/folder) to put our downloads in.

In [58]:
!mkdir /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/dependency-parsing

``cd`` into this newly made folder.

In [4]:
%cd /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/dependency-parsing

/Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/dependency-parsing


Download UDPipe.

In [44]:
!wget https://github.com/ufal/udpipe/releases/download/v1.2.0/udpipe-1.2.0-bin.zip

--2021-10-15 22:42:21--  https://github.com/ufal/udpipe/releases/download/v1.2.0/udpipe-1.2.0-bin.zip
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github-releases.githubusercontent.com/50672597/a24cacd8-77c6-11e7-8f6e-e9de8ca37f48?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211016%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211016T054221Z&X-Amz-Expires=300&X-Amz-Signature=ffd4a193d5457a90a7f862dc6c6270e379431b5fae07d1c389086d092f58477a&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=50672597&response-content-disposition=attachment%3B%20filename%3Dudpipe-1.2.0-bin.zip&response-content-type=application%2Foctet-stream [following]
--2021-10-15 22:42:21--  https://github-releases.githubusercontent.com/50672597/a24cacd8-77c6-11e7-8f6e-e9de8ca37f48?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH5

Get its contents with ``tar``.

In [45]:
!tar xvfz ./udpipe-1.2.0-bin.zip

x udpipe-1.2.0-bin/
x udpipe-1.2.0-bin/bin-linux32/
x udpipe-1.2.0-bin/bin-linux32/csharp/
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/Trainer.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/Evaluator.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/InputFormat.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/Comments.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/EmptyNodes.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/MultiwordToken.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/Sentences.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/udpipe_csharp.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/Pipeline.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/Words.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/Version.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/Children.cs
x udpipe-1.2.0-bin/bin-linux32/csharp/Ufal/UDPipe/Multiwo

``ls`` (list) the files in our current directory. Make sure we have bin-osx and ``cd`` into it.

In [5]:
!ls -l /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/dependency-parsing/udpipe-1.2.0-bin

total 1280
-rw-r--r--   1 annabellepurnomo  staff      82  2 Aug  2017 AUTHORS
-rw-r--r--   1 annabellepurnomo  staff    1555  2 Aug  2017 CHANGES
-rw-r--r--   1 annabellepurnomo  staff    3985  2 Aug  2017 INSTALL
-rw-r--r--   1 annabellepurnomo  staff   16725  2 Aug  2017 LICENSE
-rw-r--r--   1 annabellepurnomo  staff  116594  2 Aug  2017 MANUAL
-rw-r--r--   1 annabellepurnomo  staff  190921  2 Aug  2017 MANUAL.html
-rw-r--r--   1 annabellepurnomo  staff  306431  2 Aug  2017 MANUAL.pdf
-rw-r--r--   1 annabellepurnomo  staff    1309  2 Aug  2017 README
drwxr-xr-x@  5 annabellepurnomo  staff     160 15 Oct 22:43 bin-linux32
drwxr-xr-x@  5 annabellepurnomo  staff     160 15 Oct 22:43 bin-linux64
drwxr-xr-x@  6 annabellepurnomo  staff     192 16 Oct 14:37 bin-osx
drwxr-xr-x@  5 annabellepurnomo  staff     160 15 Oct 22:43 bin-win32
drwxr-xr-x@  5 annabellepurnomo  staff     160 15 Oct 22:43 bin-win64
drwxr-xr-x@ 11 annabellepurnomo  staff     352 15 Oct 22:43 bindings
drwxr-xr-x@  6 anna

In [32]:
%cd /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/dependency-parsing/udpipe-1.2.0-bin/bin-osx

/Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/dependency-parsing/udpipe-1.2.0-bin/bin-osx


Check that we have udpipe in our cur.

In [16]:
!ls

csharp                java
en_ewt-ud-test.parsed udpipe


### CoNLLU format
HEAD and DEPREL fields are used to encode a dependency tree over words

- (6th column) **HEAD**: "head" of the current word with its word index or 0

- (7th column) **DEPREL**: "universal dependency relation" to the HEAD (root if HEAD is 0)

Here is our training file in the CoNLL-U Format.
Note the part that indicates its constituents. 

In [10]:
!head /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/UD_English-EWT/en_ewt-ud-dev.conllu

# newdoc id = weblog-blogspot.com_nominations_20041117172713_ENG_20041117_172713
# sent_id = weblog-blogspot.com_nominations_20041117172713_ENG_20041117_172713-0001
# newpar id = weblog-blogspot.com_nominations_20041117172713_ENG_20041117_172713-p0001
# text = From the AP comes this story :
1	From	from	ADP	IN	_	3	case	3:case	_
2	the	the	DET	DT	Definite=Def|PronType=Art	3	det	3:det	_
3	AP	AP	PROPN	NNP	Number=Sing	4	obl	4:obl:from	_
4	comes	come	VERB	VBZ	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	0	root	0:root	_
5	this	this	DET	DT	Number=Sing|PronType=Dem	6	det	6:det	_
6	story	story	NOUN	NN	Number=Sing	4	nsubj	4:nsubj	_


Train our model ``ud_english-ewt.model`` (what we want to output) using our heldout (testing) file ``en_ewt-ud-dev.conllu`` and ``en_ewt-ud-train.conllu`` (training) file.

In [24]:
!./udpipe --train ud_english_ewt.model --heldout=/Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/UD_English-EWT/en_ewt-ud-dev.conllu /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/UD_English-EWT/en_ewt-ud-train.conllu

Loading training data: done.
Training the UDPipe model.
Training tokenizer with the following options: tokenize_url=1, allow_spaces=0, dimension=24
  epochs=100, batch_size=50, learning_rate=0.0050, dropout=0.1000, early_stopping=1
Epoch 1, logprob: -5.8952e+04, training acc: 96.04%, heldout tokens: 97.52%P/97.02%R/97.27%, sentences: 79.44%P/75.67%R/77.51%
Epoch 2, logprob: -1.2823e+04, training acc: 99.13%, heldout tokens: 98.50%P/98.11%R/98.30%, sentences: 83.42%P/78.42%R/80.84%
Epoch 3, logprob: -1.1254e+04, training acc: 99.23%, heldout tokens: 98.77%P/97.99%R/98.38%, sentences: 85.85%P/79.37%R/82.48%
Epoch 4, logprob: -9.8629e+03, training acc: 99.31%, heldout tokens: 98.86%P/98.65%R/98.75%, sentences: 87.20%P/78.92%R/82.85%
Epoch 5, logprob: -9.2433e+03, training acc: 99.33%, heldout tokens: 98.97%P/98.53%R/98.75%, sentences: 87.00%P/79.22%R/82.93%
Epoch 6, logprob: -8.7021e+03, training acc: 99.39%, heldout tokens: 98.87%P/98.77%R/98.82%, sentences: 86.64%P/80.02%R/83.20%
Epoch 

Evaluate the accuracy of your UDPipe model on test data ``en_ewt-ud-test.conllu``.

In [28]:
!./udpipe --accuracy ./ud_english_ewt.model /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/UD_English-EWT/en_ewt-ud-test.conllu


Test the accuracy of our model ``ud_english_ewt.model`` using ``en_ewt-ud-test.conllu`` as test data. 

- \-tokenize : tokenize our input
- \-tag : perform tagging on our input 
- \-parse : perform parsing on our input 
- \--input=horizontal : specifies that our input data has each sentence on a separate line, with tokens separated by spaces 

We are outputting our newly parsed file into ``en_ewt-ud-test.parsed``.

In [30]:
!./udpipe -tokenize -tag --parse --input=horizontal ud_english_ewt.model /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/UD_English-EWT/en_ewt-ud-test.txt > en_ewt-ud-test.parsed

Loading UDPipe model: done.


``cd`` back into our dependency parsing

In [38]:
%cd /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/dependency-parsing

/Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/dependency-parsing


### EVALUTATION

Download ud eval to evaluate our parser. 

In [ ]:
!wget https://universaldependencies.org/conll18/conll18_ud_eval.py

We use the ``conll18_ud_eval.py`` file we just downloaded on our gold data ``en_ewt-ud-test.conllu`` and compare it with ``en_ewt-ud-test.parsed``

In [41]:
!/Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/dependency-parsing/conll18_ud_eval.py /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/computational-tools-for-linguistic-analysis-ubc/labs/lab5_6/UD_English-EWT/en_ewt-ud-test.conllu /Users/annabellepurnomo/Desktop/LING242/LING242_assignments/LING242_assignment-2/dependency-parsing/udpipe-1.2.0-bin/bin-osx/en_ewt-ud-test.parsed 

LAS F1 Score: 74.30
MLAS Score: 63.36
BLEX Score: 65.38


Our output shows three different metrics that reflect word segmentation and the relations between content words. 

We fill focus on LAS F1 Score, which is "labeled attachment F1 score". This score is an indication of how well our words are assigned to their syntactic head and dependency label. This is a percentage of the words correctly assigned to these labels. 

**Our score of 74.30 means 74.30% of our words were correctly assigned to their syntactic head and dependency label.**